Objects/unicodeobject.c

```diff
@@ -2891,31 +2891,31 @@ unicode_from_format(_PyUnicodeWriter *writer, const char *format, va_list vargs)
                 goto fail;
         }
         else {
-            const char *p;
+            const char *p = strchr(f, '%');
             Py_ssize_t len;
-
-            p = f;
-            do
-            {
-                if ((unsigned char)*p > 127) {
-                    PyErr_Format(PyExc_ValueError,
-                        "PyUnicode_FromFormatV() expects an ASCII-encoded format "
-                        "string, got a non-ASCII byte: 0x%02x",
-                        (unsigned char)*p);
-                    goto fail;
-                }
-                p++;
+            if (p != NULL) {
+                len = p - f;
             }
-            while (*p != '\0' && *p != '%');
-            len = p - f;
-
-            if (*p == '\0')
+            else {
+                len = strlen(f);
                 writer->overallocate = 0;
+            }
 
-            if (_PyUnicodeWriter_WriteASCIIString(writer, f, len) < 0)
+            int is_ascii = (ucs1lib_find_max_char((Py_UCS1*)f, (Py_UCS1*)f + len) < 128);
```

Comment:
You can run it once, for the whole format string.

Reply:
Right. I modified my PR to run it only once, good idea!